### PR TITLE
Implement object data type with inner types.

### PR DESCRIPTION
**Pull request by seut**
Inner types of object typed columns will now be resolved and value types are ensured. This makes it possible to be complete type safe on object column values and also improves streaming.

### CODE REVIEW

File: sql/src/main/java/io/crate/metadata/information/InformationPartitionsTableInfo.java [hunk 1]
```diff
@@ -105,35 +107,57 @@ private static ColumnRegistrar createColumnRegistrar() {
...
-            .register(InformationTablesTableInfo.Columns.TABLE_SETTINGS_UNASSIGNED_NODE_LEFT_DELAYED_TIMEOUT, DataTypes.LONG)
-            .register(InformationTablesTableInfo.Columns.TABLE_SETTINGS_MAPPING, DataTypes.OBJECT)
-            .register(InformationTablesTableInfo.Columns.TABLE_SETTINGS_MAPPING_TOTAL_FIELDS, DataTypes.OBJECT)
-            .register(InformationTablesTableInfo.Columns.TABLE_SETTINGS_MAPPING_TOTAL_FIELDS_LIMIT, DataTypes.INTEGER);
+            .register(InformationTablesTableInfo.Columns.TABLE_VERSION, ObjectType.builder()
+                .setInnerType(Version.Property.CREATED.toString(), DataTypes.STRING)
+                .setInnerType(Version.Property.UPGRADED.toString(), DataTypes.STRING)
+                .build())
+            .register(InformationTablesTableInfo.Columns.TABLE_SETTINGS, ObjectType.builder()
+                //.setInnerType("refresh_interval", DataTypes.LONG)
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/metadata/information/InformationPartitionsTableInfo.java]: Can this be removed?

File: sql/src/main/java/io/crate/metadata/information/InformationPartitionsTableInfo.java [hunk 2]
```diff
@@ -105,35 +107,57 @@ private static ColumnRegistrar createColumnRegistrar() {
...
+                    .setInnerType("flush_threshold_size", DataTypes.LONG)
+                    .setInnerType("sync_interval", DataTypes.LONG)
+                    .build())
+                .setInnerType("routing", ObjectType.builder()
+                    .setInnerType("allocation", ObjectType.builder()
+                        .setInnerType("enable", DataTypes.STRING)
+                        .setInnerType("total_shards_per_node", DataTypes.INTEGER)
+                        /*
+                        .setInnerType("require", ObjectType.untyped())
+                        .setInnerType("include", ObjectType.untyped())
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/metadata/information/InformationPartitionsTableInfo.java]: Can this be removed?

File: sql/src/main/java/io/crate/metadata/information/InformationPartitionsTableInfo.java [hunk 3]
```diff
@@ -105,35 +107,57 @@ private static ColumnRegistrar createColumnRegistrar() {
...
+                        .setInnerType("exclude", ObjectType.untyped())
+                        */
+                        .build())
+                    .build())
+                .setInnerType("warmer", ObjectType.builder()
+                    .setInnerType("enabled", DataTypes.BOOLEAN)
+                    .build())
+                /*
+                .setInnerType("write", ObjectType.builder()
+                    .setInnerType("wait_for_active_shards", DataTypes.STRING)
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/metadata/information/InformationPartitionsTableInfo.java]: Can this be removed?

File: sql/src/test/java/io/crate/integrationtests/SQLTypeMappingTest.java [hunk 4]
```diff
@@ -102,6 +103,14 @@ public void setUpObjectTable() throws IOException {
         ensureYellow();
     }
 
+    /**
+     * Disabled JDBC usage cause of text mode JSON encoding which is not type safe on numeric types.
+     * E.g. byte values are always converted to integers,
+     * see {@link com.fasterxml.jackson.core.JsonGenerator#writeNumber(short)}.
+     *
+     * TODO: See how psql binary protocol behaves.
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/integrationtests/SQLTypeMappingTest.java]: I think the TODO can be removed? JSON binary is also a JSON string.

File: sql/src/test/java/io/crate/integrationtests/SQLTypeMappingTest.java [hunk 5]
```diff
@@ -259,6 +268,14 @@ public void testSetUpdate() throws Exception {
         assertEquals("127.0.0.1", response.rows()[0][11]);
     }
 
+    /**
+     * We must fix this test to run ALL statements via JDBC or not because object/map values are NOT preserving exact
+     * its elements numeric types (e.g. Long(0) becomes Integer(0)). This is caused by the usage of JSON for psql text
+     * serialization. See e.g. {@link org.codehaus.jackson.io.NumberOutput#outputLong(long, byte[], int)}.
+     *
+     * TODO: See how psql binary protocol behaves.
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/integrationtests/SQLTypeMappingTest.java]: Same here

File: sql/src/test/java/io/crate/metadata/table/TestingTableInfo.java [hunk 6]
```diff
@@ -281,6 +304,72 @@ public Builder addPartitions(String... partitionNames) {
             return this;
         }
 
+        public Builder addIndex(String name, List<String> columnNames, @Nullable String analyzerName) {
+            return addIndex(name, columnNames, analyzerName, Reference.IndexType.ANALYZED);
+        }
+
+        public Builder addIndex(String name, List<String> columnNames, @Nullable String analyzerName, Reference.IndexType indexType) {
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/metadata/table/TestingTableInfo.java]: Why are there so many extensions necessary here? They seem a bit unrelated.

(I also think we should move away from this `TestingTableInfo` stuff, rather than extending it. )

**Comment by mfussenegger** [comment]
Left some mostly minor comments. Overall looks good to me. Nice work

File: sql/src/main/java/io/crate/metadata/information/InformationPartitionsTableInfo.java (hunk 1 above)
Comment (seut) [reply sql/src/main/java/io/crate/metadata/information/InformationPartitionsTableInfo.java]: Yes, I'll create a followup task in order to match partition settings with normal table settings.

**Comment by seut** [comment]
@mfussenegger fixup pushed. I think I've found a bug related to display correct partitioned table settings, see https://github.com/crate/crate/pull/8275/files#diff-3ef1cae10e58399a881a6c7906a43b78R94. I'll investigate further and create a backport fix if needed.

**Comment by seut** [comment]
fyi: I've just verified that the change at
https://github.com/crate/crate/pull/8275/files#diff-3ef1cae10e58399a881a6c7906a43b78R94 is just testing related.
